src/Middleware/Diagnostics.EntityFrameworkCore/src/DatabaseExceptionFilter.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.Views;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore
+{
+    public class DatabaseExceptionFilter : IDeveloperPageExceptionFilter
+    {
+        private readonly ILogger _logger;
+        private readonly DatabaseErrorPageOptions _options;
+
+        public DatabaseExceptionFilter(ILogger<DatabaseExceptionFilter> logger, IOptions<DatabaseErrorPageOptions> options)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        public async Task HandleExceptionAsync(ErrorContext errorContext, Func<ErrorContext, Task> next)
+        {
+            if (errorContext.Exception is DbException)
+            {
+                try
+                {
+                    // Look for DbContext classes registered in the service provider
+                    var registeredContexts = errorContext.HttpContext.RequestServices.GetServices<DbContextOptions>()
+                        .Select(o => o.ContextType);
+
+                    if (registeredContexts.Any())
+                    {
+                        var contextDetails = new List<DatabaseContextDetails>();
+
+                        foreach (var registeredContext in registeredContexts)
+                        {
+                            var details = await errorContext.HttpContext.GetContextDetailsAsync(registeredContext, _logger);
+
+                            if (details != null)
+                            {
+                                contextDetails.Add(details);
+                            }
+                        }
+
+                        if (contextDetails.Any(c => c.PendingModelChanges || c.PendingMigrations.Any()))
+                        {
+                            var page = new DatabaseErrorPage
+                            {
+                                Model = new DatabaseErrorPageModel(errorContext.Exception, contextDetails, _options)
+                            };
+
+                            await page.ExecuteAsync(errorContext.HttpContext);
+                            return;
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    _logger.DatabaseErrorPageMiddlewareException(e);
```

Comment:
```suggestion
                    _logger.DatabaseErrorPageMiddlewareException(e);
                    return;
```

I don't think you want to fall through to the next exception handler if this one blew up. It may have failed part way through the response body.